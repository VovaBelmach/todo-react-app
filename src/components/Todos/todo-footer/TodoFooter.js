import Button from "../../UI/Button/Button";
import TodoFilter from "../todo-filter/TodoFilter";

import styles from "./TodoFooter.module.css";

const TodoFooter = (props) => {
  const todoCount = props.countTodos;
  const spanContent = todoCount === 0 ? "All done!" : `${todoCount} todo left`;

  return (
    <>
      <footer className={`${styles["todo-footer"]}`}>
        <span>{spanContent}</span>
        <TodoFilter onFilterHandler={props.onFilterHandler} />
        <Button
          onClick={props.onDeleteCompletedHandler}
          className={`${styles["todo-clear-completed"]}`}
        >
          Clear completed
        </Button>
      </footer>
    </>
  );
};

export default TodoFooter;
