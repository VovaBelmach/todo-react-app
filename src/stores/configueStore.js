import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todosSlice";

const loadStateFromLocalStorage = () => {
  const storedTodos = localStorage.getItem("todos");
  if (storedTodos) {
    return {
      todos: {
        todos: JSON.parse(storedTodos),
        filterValue: "All",
      },
    };
  }
  return undefined;
};

const persistedState = loadStateFromLocalStorage();

export default configureStore({
  reducer: {
    todos: todoReducer,
  },
  preloadedState: persistedState
});