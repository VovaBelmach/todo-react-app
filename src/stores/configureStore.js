import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todosSlice";
import { localStorageMiddleware } from "./middleware/localStorageRepositoryMiddleware"

export default configureStore({
  reducer: {
    todos: todoReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(localStorageMiddleware),
});
