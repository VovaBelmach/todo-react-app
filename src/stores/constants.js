export const TODOS_LOCAL_STORAGE_NAME = "todos";