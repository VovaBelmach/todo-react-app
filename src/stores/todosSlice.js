import { createSlice } from "@reduxjs/toolkit";
import { v4 as uuid } from "uuid";
import { saveItemsToLocalStorage } from "../repositories/localStorageRepository";
import {
  TODOS_LOCAL_STORAGE_NAME,
  TODO_FILTER_ALL_BUTTON_NAME,
  TODO_FILTER_ACTIVE_BUTTON_NAME,
  TODO_FILTER_COMPLITED_BUTTON_NAME
} from "../constants";

const initialState = {
  todos: [],
  filterValue: TODO_FILTER_ALL_BUTTON_NAME,
};

export const todosSlice = createSlice({
  name: TODOS_LOCAL_STORAGE_NAME,
  initialState: initialState,
  reducers: {
    addTodo: (state, action) => {
      const newTodo = {
        id: uuid(),
        description: action.payload.description,
        isCompleted: false,
      };
      state.todos.push(newTodo);
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
    deleteTodo: (state, action) => {
      const { id } = action.payload;
      state.todos = state.todos.filter((todo) => todo.id !== id);
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
    completeTodo: (state, action) => {
      const index = state.todos.findIndex(
        (item) => item.id === action.payload.id
      );
      if (index !== -1) {
        state.todos[index].isCompleted = !state.todos[index].isCompleted;
        saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
      }
    },
    setFilter: (state, action) => {
      state.filterValue = action.payload;
    },
    reorderTodos: (state, action) => {
      const { draggedIndex, droppedIndex } = action.payload;
      const draggedTodo = state.todos[draggedIndex];
      const newTodos = [...state.todos];
      newTodos.splice(draggedIndex, 1);
      newTodos.splice(droppedIndex, 0, draggedTodo);
      state.todos = newTodos;
      saveItemsToLocalStorage(state.todos, TODOS_LOCAL_STORAGE_NAME);
    },
  },
});

export const { addTodo, deleteTodo, completeTodo, setFilter, reorderTodos } =
  todosSlice.actions;

export default todosSlice.reducer;

export const filteredTodosSelector = (state) => {
  const { todos, filterValue } = state.todos;
  switch (filterValue) {
    case TODO_FILTER_ACTIVE_BUTTON_NAME:
      return todos.filter((todo) => todo.isCompleted !== true);
    case TODO_FILTER_COMPLITED_BUTTON_NAME:
      return todos.filter((todo) => todo.isCompleted === true);
    default:
      return todos;
  }
};
