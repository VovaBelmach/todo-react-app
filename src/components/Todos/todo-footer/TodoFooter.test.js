import React from "react";
import { Provider } from "react-redux";
import { render, screen, fireEvent } from "@testing-library/react";
import TodoFooter from "./TodoFooter";
import configureMockStore from "redux-mock-store";
import { deleteAllComlpetedTodos } from "../../../stores/todosSlice";
import {
  TODO_FOOTER_ALL_DONE_TEXT,
  TODO_FOOTER_NUMBER_TODO_LEFT_TEXT,
  TODO_FOOTER_CLEAR_COMPLETED_BUTTON_NAME,
  TODO_FILTER_ALL_BUTTON_NAME,
} from "../../../constants";

const mockStore = configureMockStore();
const store = mockStore({});

describe("TodoFooter component", () => {
  const props = {
    countTodos: 3,
    onFilterHandler: jest.fn(),
  };

  it("should match snapshot with todo count", () => {
    // Arrange
    const { container } = render(
      <Provider store={store}>
        <TodoFooter countTodos={props.countTodos} />
      </Provider>
    );

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should match snapshot with no todos", () => {
    //Arrange
    const { container } = render(
      <Provider store={store}>
        <TodoFooter countTodos={0} />
      </Provider>
    );

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render with countTodos prop", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoFooter {...props} />
      </Provider>
    );

    // Assert
    const spanElement = screen.getByText(
      `${props.countTodos} ${TODO_FOOTER_NUMBER_TODO_LEFT_TEXT}`
    );
    expect(spanElement).toBeInTheDocument();
  });

  it("should render with 'All done!' when countTodos is 0", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoFooter {...props} countTodos={0} />
      </Provider>
    );

    // Assert
    const spanElement = screen.getByText(TODO_FOOTER_ALL_DONE_TEXT);
    expect(spanElement).toBeInTheDocument();
  });

  it("should call onFilterHandler function when TodoFilter component triggers onFilterHandler event", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoFooter {...props} />
      </Provider>
    );

    // Act
    fireEvent.click(
      screen.getByRole("button", {
        name: TODO_FILTER_ALL_BUTTON_NAME,
      }),
      { target: { value: TODO_FILTER_ALL_BUTTON_NAME } }
    );

    // Assert
    expect(props.onFilterHandler).toHaveBeenCalled();
  });

  it("should call onDeleteCompletedHandler function when 'Clear completed' button is clicked", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoFooter {...props} />
      </Provider>
    );

    // Act
    const clearCompletedButton = screen.getByText(
      TODO_FOOTER_CLEAR_COMPLETED_BUTTON_NAME
    );
    fireEvent.click(clearCompletedButton);

    // Assert
    expect(store.getActions()).toContainEqual(deleteAllComlpetedTodos());
  });
});
