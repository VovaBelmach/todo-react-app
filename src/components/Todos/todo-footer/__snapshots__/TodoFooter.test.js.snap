// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter component should match snapshot with no todos 1`] = `
<footer
  class="todo-footer"
>
  <span>
    All done!
  </span>
  <ul
    class="todo-filter"
  >
    <li>
      <button
        class="button selected"
        type="button"
        value="All"
      >
        All
      </button>
    </li>
    <li>
      <button
        class="button"
        type="button"
        value="Active"
      >
        Active
      </button>
    </li>
    <li>
      <button
        class="button"
        type="button"
        value="Completed"
      >
        Completed
      </button>
    </li>
  </ul>
  <button
    class="button todo-clear-completed"
    type="button"
    value="Clear completed"
  >
    Clear completed
  </button>
</footer>
`;

exports[`TodoFooter component should match snapshot with todo count 1`] = `
<footer
  class="todo-footer"
>
  <span>
    3 todo left
  </span>
  <ul
    class="todo-filter"
  >
    <li>
      <button
        class="button selected"
        type="button"
        value="All"
      >
        All
      </button>
    </li>
    <li>
      <button
        class="button"
        type="button"
        value="Active"
      >
        Active
      </button>
    </li>
    <li>
      <button
        class="button"
        type="button"
        value="Completed"
      >
        Completed
      </button>
    </li>
  </ul>
  <button
    class="button todo-clear-completed"
    type="button"
    value="Clear completed"
  >
    Clear completed
  </button>
</footer>
`;
